src/Cassandra.Tests/SerializerTests.cs

```diff
@@ -87,61 +87,138 @@ public void EncodeDecodeSingleValuesTest()
         /// Tests that the default target type when is not provided
         /// </summary>
         [Test]
-        public void EncodeDecodeSingleValuesDefaultsFactory()
+        [TestCaseSource(nameof(SingleValuesTestCases))]
+        public void EncodeDecodeSingleValuesDefaultsFactory(object[] feed)
         {
-            var initialValues = new object[]
+            var value = ((Array)feed[0]).GetValue(0);
+            var code = (ColumnTypeCode)feed[1];
+            foreach (var protocolVersion in _protocolVersions)
+            {
+                var serializer = NewInstance(protocolVersion);
+                byte[] encoded = serializer.Serialize(value);
+                //Set object as the target CSharp type, it should get the default value
+                Assert.AreEqual(value, serializer.Deserialize(encoded, code, null));
+            }
+        }
+
+        //TODO: exhaust the types
+        static IEnumerable<object[]> SingleValuesTestCases()
+        {
+            // 2-element array, type code
+            yield return new object[] { new [] {"just utf8 text olé!", "another"}, ColumnTypeCode.Text};
+            yield return new object[] { new []{123, -1}, ColumnTypeCode.Int };
+            yield return new object[] { new []{Int64.MinValue + 100, 1}, ColumnTypeCode.Bigint };
+            yield return new object[] { new[]{-144F, 1.25F}, ColumnTypeCode.Float };
+            yield return new object[] { new []{1120D, 1.3456D}, ColumnTypeCode.Double };
+            yield return new object[] { new []{-9999.89770M, 8.0923M}, ColumnTypeCode.Decimal };
+            yield return new object[] { new[]{new DateTimeOffset(new DateTime(2010, 4, 29)), new DateTimeOffset(new DateTime(1980, 1, 9))}, ColumnTypeCode.Timestamp };
+            yield return new object[] { new[]{new IPAddress(new byte[] { 10, 0, 5, 5 }), new IPAddress(new byte[]{127,0,0,1})}, ColumnTypeCode.Inet };
+            yield return new object[] { new[]{Guid.NewGuid(), Guid.NewGuid()}, ColumnTypeCode.Uuid };
+            yield return new object[] { new[]{true, false}, ColumnTypeCode.Boolean };
+            yield return new object[] { new[]{new byte[] { 255, 128, 64, 32, 16, 9, 9 }, new byte[]{0,1,128,9,1,2,3,4}}, ColumnTypeCode.Blob };
```

Comment:
Here if I add
```java
yield return new object[] { new[] { TimeUuid.NewId(), TimeUuid.NewId() }, ColumnTypeCode.Timeuuid };
```
The tests will fail because the decoded value will be a `Guid` instead of `TimeUuid`. Is this expected?

Reply:
Hmm I believe this happens because technically the serializer is creating `Guid` objects instead of `TimeUuid` objects. The user can still retrieve them as `TimeUuid` because there's an operator defined for that conversion but `Equals` between a `TimeUuid` and a `Guid` is probably returning false even if the underlying `Guid` is the same... I'm not sure if we should make `Equals()` return true in these cases but since this is a pre existing behavior in the driver and we haven't received reports about it, I'm inclined to say that we shouldn't change `Equals()`.

For the test itself, if we want to add `TimeUuid` we'd need to add a specific assertion for that type instead of relying on the default `Equals` method.

Reply:
Ok I spent a little time debugging this because I noticed I'm using `TimeUuid` on the integration test and it is working. The issue with this test is that it is using the `Serializer` directly which bypasses some type conversion logic that we have in the `Row` class. There's some logic in `Row` that handles converting between `Guid` and `TimeUuid`. My guess is that `TimeUuid` was a C# type that was added after `Guid` was already being used for the timeuuid column type so there's some weird logic to handle converting between these two types when users ask for a `TimeUuid` to be deserialized.

So in these serializer unit tests we have to take into consideration that the `TimeUuid` serializer will return values of type `Guid`.

Reply:
Changing this line to `yield return new object[] { new[] { TimeUuid.NewId().ToGuid(), TimeUuid.NewId().ToGuid() }, ColumnTypeCode.Timeuuid };` works